### PR TITLE
Records

### PR DESCRIPTION
**Pull request by andy-morris**
- [x] Make annotation field renamable
- [x] Produce record pattern synonyms
- [x] Make the extension constructor a record if all the other constructors are
- [x] Add a record example to the documentation!!!

### CODE REVIEW

**Comment by andy-morris** [comment]
This is ready to merge I think. Unless someone wants to have a peek at how the sausage is made first.

Anyway, after this I think the version should probably be bumped to 0.2, right?

File: examples/record/Record.hs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import qualified RecordBase
+
+
+data WithString
+
+R.extendRec "Rec" [] [t|WithString|] $
+  R.defaultExtRec {
+    R.typeR1 = Ann ("label1", [t|String|]),
+    R.typeR2 = Ann ("label2", [t|String|]),
+    R.typeRecX = [("R3", "contents", [t|Int|])]
```
Comment (cwgoes) [inline examples/record/Record.hs:12]: Can the contents also have record fields, or not?

File: src/Extensible.hs [hunk 2]
```diff
@@ -424,32 +504,47 @@ strict = Bang NoSourceUnpackedness SourceStrict
 appExtTvs :: TH.Type -> Name -> [TyVarBndr] -> TH.Type
 appExtTvs t ext tvs = foldl AppT t $ fmap VarT $ ext : fmap tyvarName tvs
 
--- | Generate an extended constructor by renaming it and replacing recursive
--- occrences of the datatype.
+-- | Generate an extended constructor by renaming it, replacing recursive
+-- occrences of the datatype, and adding an extension field at the end
```
Comment (cwgoes) [inline src/Extensible.hs]:
```suggestion
-- occurrences of the datatype, and adding an extension field at the end
```

**Review by cwgoes — APPROVED** [review]
ACK, one typo & one question

File: examples/record/Record.hs (hunk 1 above)
Comment (andy-morris) [reply examples/record/Record.hs:12]: I'm not sure what you mean, sorry

In this case the extension constructor comes out to `R3 {contents :: Int}`

Comment (cwgoes) [reply examples/record/Record.hs:12]: Ah, I see - can there be multiple fields (by using a 5-tuple, I suppose, for two fields)?

**Pull request merged**